test/functional/lcow_uvm_test.go

```diff
@@ -0,0 +1,430 @@
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = false
+				opts.KernelFile = uvm.KernelFile
+			},
+			wantArgs: []string{`root=/dev/pmem0`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/sda`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+		{
+			name: "SCSI no vPMEM VHD kernel",
+			optsFn: func(opts *uvm.OptionsLCOW) {
+				opts.SCSIControllerCount = 1
+				opts.VPMemDeviceCount = 0
+
+				opts.PreferredRootFSType = uvm.PreferredRootFSTypeVHD
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = false
+				opts.KernelFile = uvm.KernelFile
+			},
+			wantArgs: []string{`root=/dev/sda`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/pmem`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+		{
+			name: "no SCSI single vPMEM VHD vmlinux",
+			optsFn: func(opts *uvm.OptionsLCOW) {
+				opts.SCSIControllerCount = 0
+				opts.VPMemDeviceCount = 1
+
+				opts.PreferredRootFSType = uvm.PreferredRootFSTypeVHD
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = true
+				opts.KernelFile = uvm.UncompressedKernelFile
+			},
+			wantArgs: []string{`root=/dev/pmem0`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/sda`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+		{
+			name: "SCSI no vPMEM VHD vmlinux",
+			optsFn: func(opts *uvm.OptionsLCOW) {
+				opts.SCSIControllerCount = 1
+				opts.VPMemDeviceCount = 0
+
+				opts.PreferredRootFSType = uvm.PreferredRootFSTypeVHD
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = true
+				opts.KernelFile = uvm.UncompressedKernelFile
+			},
+			wantArgs: []string{`root=/dev/sda`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/pmem`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			opts := defaultLCOWOptions(ctx, t)
+			opts.ProcessorCount = numCPU
+			tc.optsFn(opts)
+
+			if opts.KernelDirect {
+				require.Build(t, 18286)
+			}
+
+			vm := testuvm.CreateAndStartLCOWFromOpts(ctx, t, opts)
+
+			// validate the kernel args were constructed as expected
+			ioArgs := testcmd.NewBufferedIO()
+			cmdArgs := testcmd.Create(ctx, t, vm, &specs.Process{Args: []string{"cat", "/proc/cmdline"}}, ioArgs)
+			testcmd.Start(ctx, t, cmdArgs)
+			testcmd.WaitExitCode(ctx, t, cmdArgs, 0)
+
+			ioArgs.TestStdOutContains(t, tc.wantArgs, tc.notWantArgs)
+
+			// some boot options (notably using initrd) need to validated by looking at dmesg logs
```

Comment:
If `dmesg` output has the entire command line (i.e. what we are reading from `/proc/cmdlind` plus some other options) why not only read dmesg logs and avoid reading `/proc/cmdline`?

Reply:
basically cause the `TestStdOutContains` function checks that the sting contains each of args, so unless we first parse the dmesg output for only the command line bit, the check might turn a false positive if the arg we want shows up elsewhere in the logs.
I didn't want to assume a stable output for dmesg, nor a stable ordering of the arguments in the command line (eg, searching for the string `root=/dev/sda rootwait init=/init` instead of each component individually)